test/e2e/storage/host_path_type.go

```diff
@@ -0,0 +1,435 @@
+	})
+
+	ginkgo.It("Should be able to mount block device 'ablkdev' successfully when HostPathType is HostPathBlockDev", func() {
+		verifyPodHostPathType(f, basePod.Spec.NodeName, targetBlockDev, &hostPathBlockDev)
+	})
+
+	ginkgo.It("Should be able to mount block device 'ablkdev' successfully when HostPathType is HostPathUnset", func() {
+		verifyPodHostPathType(f, basePod.Spec.NodeName, targetBlockDev, &hostPathUnset)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathDirectory", func() {
+		verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, targetBlockDev, fmt.Sprintf("%s is not a directory", targetBlockDev), &hostPathDirectory)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathFile", func() {
+		verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, targetBlockDev, fmt.Sprintf("%s is not a file", targetBlockDev), &hostPathFile)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathSocket", func() {
+		verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, targetBlockDev, fmt.Sprintf("%s is not a socket", targetBlockDev), &hostPathSocket)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathCharDev", func() {
+		verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, targetBlockDev, fmt.Sprintf("%s is not a character device", targetBlockDev), &hostPathCharDev)
+	})
+})
+
+func newHostPathTypeTestPod(nodeName, hostDir, mountDir string, hostPathType *v1.HostPathType) *v1.Pod {
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "test-hostpath-type-",
+		},
+		Spec: v1.PodSpec{
+			NodeName:      nodeName,
+			RestartPolicy: v1.RestartPolicyNever,
+			Containers: []v1.Container{
+				{
+					Name:  "host-path-testing",
+					Image: imageutils.GetE2EImage(imageutils.Agnhost),
+					VolumeMounts: []v1.VolumeMount{
+						{
+							Name:      "host",
+							MountPath: mountDir,
+							ReadOnly:  false,
+						},
+					},
+				},
+			},
+			Volumes: []v1.Volume{
+				{
+					Name: "host",
+					VolumeSource: v1.VolumeSource{
+						HostPath: &v1.HostPathVolumeSource{
+							Path: hostDir,
+							Type: hostPathType,
+						},
+					},
+				},
+			},
+		},
+	}
+	return pod
+}
+
+func verifyPodHostPathTypeFailure(f *framework.Framework, nodeName, hostDir, pattern string, hostPathType *v1.HostPathType) {
+	pod := newHostPathTypeTestPod(nodeName, hostDir, "/mnt/test", hostPathType)
+	ginkgo.By(fmt.Sprintf("Creating pod %s", pod.Name))
+	pod, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Create(context.TODO(), pod, metav1.CreateOptions{})
+	framework.ExpectNoError(err)
+
+	err = e2epod.WaitTimeoutForPodRunningInNamespace(f.ClientSet, pod.Name, pod.Namespace, framework.PodStartShortTimeout)
+	framework.ExpectError(err)
+
+	ginkgo.By("Checking for HostPathType error event")
```

Comment:
We also had flakiness issues in the past where tests were checking for events. However events could be dropped if there's too many events in the system.

Is it possible to check for Pod status/condition instead?

Reply:
> Is it possible to check for Pod status/condition instead?

Seems not. Currently related mount failure are only emitted by "FailedMount" events.

> However events could be dropped if there's too many events in the system.

I don't think this will be a big deal for this test case, since the events will be only within 1~2 minutes if needed.

Reply:
Unfortunately this has been a source of flakiness in the past. What we've done for other tests in the past, is check for events to speed up the test, but if we can't find an event after some time, we don't fail the test.

See https://github.com/kubernetes/kubernetes/blob/b9c370c90f71d4a6a5ad5d4668de6d3fb987720c/test/e2e/storage/testsuites/volumemode.go#L240 for an example